proposals/build-provenance-and-code-signing-for-homebrew.md

```diff
@@ -0,0 +1,112 @@
+    - In the interest of simplicity, this proposal limits tap identities to just GitHub Actions identities, as these are "intrinsic" to the tap itself and therefore require no additional user trust decisions. Future extensions to this scheme may extend tap identities to include email, user, or other identity types supported by Sigstore.
+- A Sigstore client library capable of verifying Sigstore signatures.
+
+Client-side verification of bottles requires `brew` to know which bottles are actually signed. Clients should use the "flag day" metadata above to establish this: `brew install` should refuse to install any unsigned bottle from a tap that both had a flag day _and_ whose flag day is before the local system time.
+
+The high-level flow for bottle signature verification, at tap time:
+
+- A user performs either an explicit tap (`brew tap foo/bar`) or an implicit tap (`brew install foo/bar/baz`), and the local `brew` client establishes a trust proposition based on that action ("bottles from `foo/bar` must be signed by the `foo/bar` identity").
+- The local `brew` client looks for the "flag day" metadata and, if present and before the local system time, enforces signature verification for all bottles referenced by the tap.
+
+And at `brew install` time:
+
+- For each bottle resolved during a `brew install` invocation, the bottle must have an accompanying signature. If the accompanying signature is not present, then `brew install` must fail.
+- For each bottle and signature, `brew install` must internally reconstruct the _attestation_ described in the signing process above. This is then used as the cleartext input for signature verification. This input is then passed into the underlying Sigstore client along with its associated signature; the Sigstore client is responsible for performing both the signature verification and its associated identity verification.
+- If the underlying signature verification fails, then `brew install` must fail.
+
+Like with previous Homebrew features, an experimental feature-flagged variant of signature verification may be merged and enabled before it is enabled by default. Verification will be optional while bottles are incrementally updated/rebuilt with signatures, and once every core bottle is signed, then everything must be signed after the flag day.
+
+## Full source and build provenance for Homebrew
+
+Bottle signing and subsequent verification unlocks a potential subsequent improvement: because Homebrew performs source builds to produce bottles, Homebrew could also verify signatures for those source artifacts. That signature could then be _countersigned_ for by Homebrew's own bottle signatures, effectively producing a strong cryptographic association from the original source artifact (produced by a third-party upstream) to the Homebrew-produced binary artifact.
```

Comment:
The idea is that the source (say the tarball of ripgrep from the example below) is signed by its author and brew install would verify its signature as well as the provenance for the build signed by homebrew-core's identity.

I'm trying to understand the value of the signature on the source tarball. In option 1, the signature is expected to be from the identity associated with the github repo. Is the ability to issue this signature for the tarball more protected than the capability to upload a release? Does it require 2FA every time, for example?

Option 2 is more interesting because the upstream maintainer for a package can be associated with a separate, non-github identity, but I'm also curious about the overhead involved. Does it fall to the homebrew maintainers to verify that the github repo and the secondary ID are owned by the same person? Also, how is the other identity updated / rotated / revoked? Right now it seems like all of this is manual.

Finally, apart from verifying a signature on the source and verifying the built artifact's provenance / sig, the two must also be cross checked to ensure the source signed in the first case is identical to what's used as the input to the build. As @trishankatdatadog notes above, in-toto layouts help here.

Reply:
I'll add that TUF would be a good option for doing the association for option 2. An offline delegation to a separate developer identity would provide additional security, and provide an easy way for clients to determine which identities to use for verification.

Reply:
Note that developer identity is out of scope currently for the MVP, that the identity provided by GitHub Actions and is linked to source will be used.

Reply:
@haydentherapper, so the signature on the source tarball must be issued by github actions on that repository? As in, the release must be automated using github actions?

Reply:
Correct, given all core bottles are currently built using GitHub Actions, we can construct an identity that represents the GitHub Action. https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md, in particular https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md#1361415726419--build-signer-uri, has more information.

See https://github.com/ossf/wg-securing-software-repos/commit/0f9675a427973d84ee56f84c821b12dd526c987c

Reply:
Yeah, this section was written with machine identity in mind: it assumes something like GitHub Actions (or another build/CI identity supported by Sigstore), which means that verifying and then countersigning the upstream source's signature here would include embedded claims about the source provenance (e.g. the source URL, which workflow built it, etc.).

>  the signature is expected to be from the identity associated with the github repo. Is the ability to issue this signature for the tarball more protected than the capability to upload a release? Does it require 2FA every time, for example?

The ability to create a verifiable machine identity signature on GitHub Actions is gated behind a workflow permission, which is comparable (in practice) to having write access to the repository. So it's *similar* to having the ability to make a release; the value comes from it being verifiably bound to the identity that made the release.

Reply:
May I then recommend refraining from talking about Source altogether for the MVP? Because I think the problem is trickier to solve for right now than it looks like. We can certainly save it for a future iteration 🙂

Reply:
Works for me -- I think this is technically outside of the scope anyways (since we're focusing on build provenance for this proposal and not source provenance), so if @di is okay with an "out of scope" note I'll add it.

Reply:
FWIW, source in this case != source provenance. There's a link between source and build because of the values in the GHA identity. It's accurate to say we're not tackling source provenance, but being able to link the identity to source to build is what makes verification seamless for this.

Reply:
So, if I understand @haydentherapper and @woodruffw correctly, the idea is that the _provenance_ attestation cryptographically binds the source code repository (and the build platform) to the artifact, correct?

Reply:
Correct, nothing about the legitimacy of the source code. Concretely, here's an example of an identity: https://github.com/slsa-framework/slsa-github-generator/.github/workflows/generator_container_slsa3.yml@v1.4.0

This includes the source code repository and the versioned builder (reusable workflow) that creates the artifact.